sdk/schemaregistry/azure-schemaregistry-client/src/main/java/com/azure/schemaregistry/client/CachedSchemaRegistryClient.java

```diff
@@ -0,0 +1,281 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.schemaregistry.client;
+
+import com.azure.core.annotation.ServiceClient;
+import com.azure.core.exception.HttpResponseException;
+import com.azure.core.http.HttpPipeline;
+import com.azure.core.util.logging.ClientLogger;
+import com.azure.schemaregistry.client.rest.AzureSchemaRegistryRestService;
+import com.azure.schemaregistry.client.rest.AzureSchemaRegistryRestServiceClientBuilder;
+import com.azure.schemaregistry.client.rest.models.GetSchemaByIdResponse;
+import com.azure.schemaregistry.client.rest.models.SchemaId;
+
+import java.nio.charset.Charset;
+import java.nio.charset.StandardCharsets;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Locale;
+import java.util.UUID;
+import java.util.function.Function;
+
+/**
+ * HTTP-based client that interacts with Azure Schema Registry service to store and retrieve schemas on demand.
+ * <p>
+ * Utilizes in-memory HashMap caching to minimize network I/O.
+ * <p>
+ * Max HashMap size can be configured when instantiating.
+ * Two maps are maintained -
+ * - SchemaRegistryObject cache by GUID - stores GUIDs previously seen in payloads
+ * - SchemaRegistryObject cache by schema string - minimizes HTTP calls when sending payloads of same schema
+ * <p>
+ * TODO: implement max age for schema maps? or will schemas always be immutable?
+ *
+ * @see SchemaRegistryClient Implements SchemaRegistryClient interface to allow for testing with mock
+ * @see CachedSchemaRegistryClientBuilder Follows builder pattern for object instantiation
+ */
+@ServiceClient(
+    builder = CachedSchemaRegistryClientBuilder.class,
+    serviceInterfaces = AzureSchemaRegistryRestService.class)
+public class CachedSchemaRegistryClient implements SchemaRegistryClient {
+    private final ClientLogger logger = new ClientLogger(CachedSchemaRegistryClient.class);
+
+    public static final int MAX_SCHEMA_MAP_SIZE_DEFAULT = 1000;
+    public static final Charset SCHEMA_REGISTRY_SERVICE_ENCODING = StandardCharsets.UTF_8;
+    static final int MAX_SCHEMA_MAP_SIZE_MINIMUM = 10;
+
+    private final AzureSchemaRegistryRestService restService;
+    private final int maxSchemaMapSize;
+    private final HashMap<String, Function<String, Object>> typeParserDictionary;
```

Comment:
I see that `typeParserDictionary` generally has it's key lower cased to allow for case insensitive matching, I believe using a `TreeMap` should cover that case without needing to manually manage key casing on accesses. Using `new TreeMap(String::equalsIgnoreCase)` should produce the same effect.

https://docs.oracle.com/javase/8/docs/api/java/util/TreeMap.html#TreeMap-java.util.Comparator-

Reply:
Using ConcurrentSkipListMap<>(String.CASE_INSENSITIVE_ORDER)